File: src/main/java/ru/ensemplix/command/CommandDispatcher.java
package ru.ensemplix.command;

import com.google.common.base.Throwables;
import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;

import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.lang.reflect.ParameterizedType;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Основной класс для работы с командами. Здесь регистрируются команды,
 * парсеры и выполняются команды.
 */
public class CommandDispatcher {

    /**
     * Список всех команд.
     */
    private final Multimap<String, CommandHandler> commands = ArrayListMultimap.create();

    /**
     * Список всех парсеров.
     */
    private final Map<Class, TypeParser> parsers = new HashMap<>();

    /**
     * Нужно ли убирать первый символ("/", "!", "@") при выполнении команды.
     * По умолчанию символ убирается.
     */
    private final boolean removeFirstChar;

    public CommandDispatcher() {
        this(true);
    }

    public CommandDispatcher(boolean removeFirstChar) {
        this.removeFirstChar = removeFirstChar;

        // Стандартные парсеры библиотеки.
        bind(String.class, new StringParser());
        bind(Integer.class, new IntegerParser());
        bind(int.class, new IntegerParser());
        bind(Boolean.class, new BooleanParser());
        bind(boolean.class, new BooleanParser());
        bind(Float.class, new FloatParser());
        bind(float.class, new FloatParser());
        bind(Double.class, new DoubleParser());
        bind(double.class, new DoubleParser());
    }

    /**
     * Выполнении команды, отправленной пользователем. Если команда не будет найдена, то
     * будет брошено исключение CommandNotFoundException. Возвращаемый результат зависит
     * от результата выполнения команды и может использоваться для логирования.
     */
    public boolean call(CommandSender sender, String cmd) throws CommandNotFoundException {
        checkNotNull(sender, "Please provide command sender");
        checkNotNull(cmd, "Please provide command line");
        checkArgument(cmd.length() > 1, "Please provide valid command line");

        if(removeFirstChar) {
            cmd = cmd.substring(1);
        }

        String[] args = cmd.split(" ");
        CommandHandler command = getCommand(args[0], null);
        int argsFrom = 0;

        if(args.length > 1) {
            CommandHandler subcommand = getCommand(args[0], args[1]);
            argsFrom = 1;

            if(subcommand != null) {
                command = subcommand;
            }
        }

        // Если команда не найдена, то выбрасываем исключение.
        if(command == null) {
            throw new CommandNotFoundException();
        }

        Method method = command.getMethod();
        Parameter[] parameters = method.getParameters();
        int length = parameters.length;

        Object[] parsed = new Object[length];
        parsed[0] = sender;

        // Подготавливаем аргументы команды.
        for (int i = 1; i < length; i++) {
            int pos = argsFrom + i;

            if(Iterable.class.isAssignableFrom(parameters[i].getType())) {
                ParameterizedType type = (ParameterizedType) parameters[i].getParameterizedType();
                TypeParser parser = parsers.get(type.getActualTypeArguments()[0]);
                Collection<Object> collection = new LinkedList<>();

                for(int y = pos; y < args.length; y++) {
                    collection.add(parser.parse(args[y]));
                }

                parsed[i] = collection;
            } else {
                TypeParser parser = parsers.get(parameters[i].getType());

                if (args.length - 1 >= pos) {
                    parsed[i] = parser.parse(args[pos]);
                } else {
                    parsed[i] = parser.parse(null);
                }
            }
        }

        // Выполняем команду.
        try {
            Object result = method.invoke(command.getObject(), parsed);
            return result == null || (boolean) result;
        } catch (Exception e) {
            Throwables.propagate(e);
        }

        return false;
    }

    /**
     * Регистрация команды происходит по любому объекту с помощью аннотации @Command.
     * Количество имен команды на ограничено.
     */
    public void register(Object obj, String... names) {
        checkNotNull(obj, "Please provide valid command");
        checkNotNull(names, "Please provide valid command name");

        // Проверяем, что команды с таким именем еще нет.
        for(String name : names) {
            if(getCommand(name, null) != null) {
                throw new IllegalArgumentException("Command with name " + name + " already exists");
            }
        }

        for (Method method : obj.getClass().getMethods()) {
            Command annotation = method.getAnnotation(Command.class);

            // Команда должна обязательно быть помечена аннотацией @Command.
            if(annotation == null) {
                continue;
            }

            // Команда должна обязательно возвращать void или boolean.
            if(method.getReturnType() != void.class && method.getReturnType() != boolean.class) {
                throw new IllegalArgumentException(method + " must return void or boolean");
            }

            Parameter[] parameters = method.getParameters();
            int length = parameters.length;

            // Обязательно первым параметром команды должен быть ее отправитель.
            if(length == 0 || !CommandSender.class.isAssignableFrom(parameters[0].getType())) {
                throw new IllegalArgumentException("Please provide command sender for " + method);
            }

            // Проверяем что все параметры команды будут отработаны корректно.
            for(int i = 1; i < length; i++) {
                if(Iterable.class.isAssignableFrom(parameters[i].getType())) {
                    if(i + 1 != length) {
                        throw new IllegalArgumentException("Iterable must be last parameter in " + method);
                    }
                } else if(!parsers.containsKey(parameters[i].getType())) {
                    throw new IllegalArgumentException("Please provide type parser for " + parameters[i].getType());
                }
            }

            CommandHandler handler = new CommandHandler(names[0], method.getName(), method, obj);

            for(String name : names) {
                commands.put(name, handler);
            }
        }
    }

    /**
     * Регистрация парсера для конвертации строки в объект.
     */
    public void bind(Class<?> clz, TypeParser parser) {
        parsers.put(clz, parser);
    }

    private CommandHandler getCommand(String command, String subCommand) {
        for (CommandHandler handler : commands.get(command)) {
            if(subCommand == null) {
                if(handler.getOrigin().equals(handler.getName())) {
                    return handler;
                }
            } else if (handler.getName().equals(subCommand)) {
                return handler;
            }
        }

        return null;
    }

    private class StringParser implements TypeParser<String> {
        @Override
        public String parse(String value) {
            return value;
        }
    }

    private class IntegerParser implements TypeParser<Integer> {
        @Override
        public Integer parse(String value) {
            try {
                return Integer.parseInt(value);
            } catch(NumberFormatException e) {
                return 0;
            }
        }
    }

    private class BooleanParser implements TypeParser<Boolean> {
        @Override
        public Boolean parse(String value) {
            try {
                return Boolean.parseBoolean(value);
            } catch(NumberFormatException e) {
                return false;
            }
        }
    }

    private class FloatParser implements TypeParser<Float> {
        @Override
        public Float parse(String value) {
            try {
                return Float.parseFloat(value);
            } catch(NumberFormatException e) {
                return 0F;
            }
        }
    }

    private class DoubleParser implements TypeParser<Double> {
        @Override
        public Double parse(String value) {
            try {
                return Double.parseDouble(value);
            } catch(NumberFormatException e) {
                return 0D;
            }
        }
    }

}

File: src/main/kotlin/ru/ensemplix/command/CommandDispatcher.kt
package ru.ensemplix.command
import ru.ensemplix.command.argument.Argument
import ru.ensemplix.command.argument.ArgumentParser
import ru.ensemplix.command.argument.ArgumentParser.*
import ru.ensemplix.command.exception.CommandAccessException
import ru.ensemplix.command.exception.CommandException
import ru.ensemplix.command.exception.CommandNotFoundException
import java.lang.reflect.ParameterizedType
import java.util.*

/**
 * Основной класс для работы с командами.
 */
open class CommandDispatcher {

    companion object {
        /**
         * Список команд.
         */
        @JvmField
        val commands = HashMap<String, CommandHandler>()

        /**
         * Список автомачических дополнений команды.
         */
        @JvmField
        val completers = HashMap<Class<*>, CommandCompleter>()

        /**
         * Список парсеров в объекты.
         */
        @JvmField
        val parsers = HashMap<Class<*>, ArgumentParser<*>>()
    }

    init {
        // Примитивные парсеры.
        bind(String::class.java, StringArgumentParser())
        bind(Int::class.java, IntegerArgumentParser())
        bind(Integer.TYPE, IntegerArgumentParser())
        bind(Boolean::class.java, BooleanArgumentParser())
        bind(java.lang.Boolean.TYPE, BooleanArgumentParser())
        bind(Float::class.java, FloatArgumentParser())
        bind(java.lang.Float.TYPE, FloatArgumentParser())
        bind(Double::class.java, StringArgumentParser())
        bind(java.lang.Double.TYPE, StringArgumentParser())
    }

    /**
     * Выполнение команды, отправленной пользователем, на основе отправленного текста.
     *
     * Если команда не существует или нет такого действия, то будет выброшено исключение.
     * {@link CommandNotFoundException} CommandNotFoundException.
     *
     * Если пользователю нельзя выполнять указанную команду, то будет выброшено
     * исключение {@link CommandAccessException} CommandAccessException.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return {@code true}, если команда была выполнена без ошибок.
     * @throws CommandException Выбрасывает исключение, если команды не
     * существует или нет разрешения на ее выполнение.
     */
    @Throws(CommandException::class)
    fun call(sender: CommandSender, cmd: String): CommandResult {
        val context = validate(sender, cmd)
        val action = context.action

        if(action == null) {
            throw CommandNotFoundException()
        }

        val method = action.method
        val args = context.args
        val parameters = method.parameters
        val length = parameters.size
        val arguments = ArrayList<Argument<*>>()
        val parsed = arrayOfNulls<Any>(length)
        parsed[0] = sender

        for(i in 1..length - 1) {
            val parameterType = parameters[i].type
            val parser: ArgumentParser<*>?

            if(Iterable::class.java.isAssignableFrom(parameterType)) {
                val type = parameters[i].parameterizedType as ParameterizedType
                val argumentType = type.actualTypeArguments[0]

                if(argumentType is ParameterizedType && Argument::class.java.isAssignableFrom(argumentType.rawType as Class<*>)) {
                    parser = parsers[argumentType.actualTypeArguments[0]]
                } else {
                    parser = parsers[argumentType]
                }
            } else if(Argument::class.java.isAssignableFrom(parameterType)) {
                val type = parameters[i].parameterizedType as ParameterizedType
                parser = parsers[type.actualTypeArguments[0]]
            } else {
                parser = parsers[parameterType]
            }

            if(Iterable::class.java.isAssignableFrom(parameterType)) {
                // Подготоваливаем коллекцию.
                val collection = ArrayList<Any?>()

                for(y in i - 1..args.size - 1) {
                    val type = parameters[i].parameterizedType as ParameterizedType
                    val argumentType = type.actualTypeArguments[0]
                    val argument = parser!!.parseArgument(args[y])

                    if(argumentType is ParameterizedType && Argument::class.java.isAssignableFrom(argumentType.rawType as Class<*>)) {
                        collection.add(argument)
                    } else {
                        collection.add(argument.value)
                    }

                    if(argument.text == null) {
                        argument.text = args[y]
                    }

                    arguments.add(argument)
                }
                parsed[i] = collection
            } else {
                // Подготавливаем аргументы команды.
                val argument: Argument<*>
                if(args.size + 1 > i) {
                    argument = parser!!.parseArgument(args[i - 1])
                    if(argument.text == null) {
                        argument.text = args[i - 1]
                    }
                } else {
                    argument = parser!!.parseArgument(null)
                }

                if(Argument::class.java.isAssignableFrom(parameterType)) {
                    parsed[i] = argument
                } else {
                    parsed[i] = argument.value
                }

                arguments.add(argument)
            }
        }
        // Выполняем команду.
        try {
            // Если возвращает void, то считаем что результат выполнения команды всегда положительный.
            val result = method.invoke(context.handler.obj, *parsed)
            val success = result == null || result as Boolean
            return CommandResult(context, arguments, success)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }
    /**
     * Автоматическое дополнение команды на основе ввода пользователя.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return Возвращает список возможных вариантов автодополнения.
     */
    fun complete(sender: CommandSender, cmd: String): Collection<String> {
        val context: CommandContext

        try {
            context = validate(sender, cmd)
        } catch (e: CommandException) {
            val matches = ArrayList<String>()
            val names = commands.keys

            if(cmd.length > 0) {
                names.forEach {
                    if(it.startsWith(cmd)) {
                        matches.add("/$it")
                    }
                }

                return matches
            }

            names.forEach { matches.add("/$it") }
            return matches
        }

        val actions = context.handler.actions.keys
        val action = context.action
        val args = context.args

        if(args.size == 0 && cmd.last() != ' ') {
            return emptyList<String>()
        }

        if(args.size == 1 && (action == null || context.handler.main != null)) {
            val matches = ArrayList<String>()

            actions.forEach {
                if(it.startsWith(args[0])) {
                    matches.add(it)
                }
            }

            if(matches.isNotEmpty()) {
                return matches
            }
        }

        if(action == null && context.handler.main == null) {
            return actions
        }

        var arg = ""
        var i = 1

        if(args.size > 0) {
            i = args.size
            arg = args[i - 1]
        }

        val parameters = context.action!!.method.parameters
        val parameterType = parameters[i].type
        val completer: CommandCompleter?

        if(!Iterable::class.java.isAssignableFrom(parameterType)) {
            val argsLength = if(cmd.last() == ' ') args.size + 1 else args.size

            if(argsLength > context.action.method.parameterCount - 1) {
                return emptyList<String>()
            }
        }

        if(Iterable::class.java.isAssignableFrom(parameterType)) {
            val type = parameters[i].parameterizedType as ParameterizedType
            val argumentType = type.actualTypeArguments[0]

            if(argumentType is ParameterizedType && Argument::class.java.isAssignableFrom(argumentType.rawType as Class<*>)) {
                completer = completers[argumentType.actualTypeArguments[0]]
            } else {
                completer = completers[argumentType]
            }
        } else if(Argument::class.java.isAssignableFrom(parameterType)) {
            val type = parameters[i].parameterizedType as ParameterizedType
            completer = completers[type.actualTypeArguments[0]]
        } else {
            completer = completers[parameterType]
        }

        if(completer != null) {
            return completer.complete(context, arg)
        }

        return emptyList<String>()
    }

    /**
     * Проверяет строку и конвертирует результат проверки в объект
     * {@link CommandContext} CommandContext.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return Возвращает результат проверки.
     * @throws CommandException Выбрасывает исключение, если команды не
     * существует или нет разрещения на ее выполнение.
     */
    @Throws(CommandException::class)
    private fun validate(sender: CommandSender, cmd: String?): CommandContext {
        if(cmd == null || cmd.isEmpty()) {
            throw CommandNotFoundException()
        }

        var args = cmd.toLowerCase().split((" ").toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
        val handler = commands[args[0]]

        if(handler == null) {
            throw CommandNotFoundException()
        }

        val commandActions = handler.actions
        var action: CommandAction? = null

        if(args.size > 1 && commandActions.containsKey(args[1])) {
            val actions = commandActions[args[1]]!!
            args = Arrays.copyOfRange<String>(args, 2, args.size)

            if(actions.size == 1) {
                action = actions[0]
            } else {
                for(possibleAction in actions) {
                    if(possibleAction.method.parameterCount - 1 >= args.size) {
                        if(action != null && possibleAction.method.parameterCount > action.method.parameterCount) {
                            continue
                        }

                        action = possibleAction
                    }
                }
            }
        } else {
            args = Arrays.copyOfRange<String>(args, 1, args.size)

            if(handler.main != null) {
                action = handler.main
            }
        }

        val actionName: String? = if(action != null) action.method.name.toLowerCase() else null

        if(action != null) {
            val main = handler.main
            val method = action.method
            val checkAction = if(main == null || main.method != method) method.name.toLowerCase() else null

            if(action.annotation.permission && !sender.canUseCommand(handler.name, checkAction)) {
                throw CommandAccessException()
            }
        }

        return CommandContext(handler.name, actionName, action, args, handler)
    }
    /**
     * Регистрация команды происходит по методам, которые содержат аннотацию
     * {@link Command} @Command. Количество имен для команды неограничено.
     * Обязательно должна быть хотя бы одна команда.
     *
     * @param obj Объект, в котором мы ищем команды.
     * @param names Названия команд.
     */
    fun register(obj: Any, vararg names: String?) {
        // Проверяем, что команды с таким именем еще нет.
        for(name in names) {
            if(name == null || name.isEmpty()) {
                throw IllegalArgumentException("Please provide valid command name")
            }

            if(name.contains(' ')) {
                throw IllegalArgumentException("Please provide command name with no whitespace")
            }

            if(commands.containsKey(name.toLowerCase())) {
                throw IllegalArgumentException("Command with name " + name.toLowerCase() + " already exists")
            }
        }

        val commandActions = HashMap<String, ArrayList<CommandAction>>()
        var main: CommandAction? = null

        for(method in obj.javaClass.methods) {
            val annotation = method.getAnnotation(Command::class.java)

            // Команда обязательно должна быть помечена аннотацией @Command.
            if(annotation == null) {
                continue
            }

            // Команда должна обязательно возвращать void или boolean.
            if(method.getReturnType() != Void.TYPE && method.getReturnType() != java.lang.Boolean.TYPE) {
                throw IllegalArgumentException(method.getName() + " must return void or boolean")
            }

            val parameters = method.parameters
            val length = parameters.size

            // Первым параметром команды обязательно должен быть ее отправитель.
            if(length == 0 || !CommandSender::class.java.isAssignableFrom(parameters[0].type)) {
                throw IllegalArgumentException("Please provide command sender for " + method.name)
            }

            // Проверяем, что все параметры команды будут отработаны корректно.
            for(i in 1..length - 1)  {
                var parameterType = parameters[i].type

                if(Iterable::class.java.isAssignableFrom(parameterType)) {
                    if(i + 1 != length) {
                        throw IllegalArgumentException("Iterable must be last parameter in " + method.name)
                    }
                } else {
                    if(Argument::class.java.isAssignableFrom(parameterType)) {
                        val type = parameters[i].parameterizedType as ParameterizedType
                        parameterType = type.actualTypeArguments[0] as Class<*>
                    }

                    if(!parsers.containsKey(parameterType)) {
                        throw IllegalArgumentException("Please provide type parser for " + parameterType)
                    }
                }
            }

            val action = CommandAction(method, annotation)

            if(annotation.main) {
                main = action
            }

            val methodName = method.name.toLowerCase()

            if(!commandActions.containsKey(methodName)) {
                commandActions.put(methodName, ArrayList<CommandAction>())
            }

            commandActions.get(methodName)!!.add(action)
        }

        if(commandActions.isEmpty()) {
            throw IllegalStateException("Not found any method marked with @Command")
        }

        for(name in names) {
            commands[name!!.toLowerCase()] = CommandHandler(names[0]!!.toLowerCase(), obj, main, commandActions)
        }
    }

    /**
     * Удаляет все команды, связанные с выбранным классом.
     *
     * @param cls Класс, который мы удаляем из команд.
     */
    fun unregister(cls: Class<*>) {
        val iterator = commands.values.iterator()

        while(iterator.hasNext()) {
            if(iterator.next().obj.javaClass.equals(cls)) {
                iterator.remove()
            }
        }
    }

    /**
     * Регистрация парсера для конвертации строки в объект.
     *
     * @param clz Класс, который мы будем конвертировать в объект.
     * @param parser Парсер, который знает как парсить класс.
     */
    fun bind(clz: Class<*>, parser: ArgumentParser<*>) {
        parsers[clz] = parser
    }

    /**
     * Регистрация дополнителя для автодополнения команды.
     *
     * @param clz Класс, который мы будем автодополнять.
     * @param completer Дополнитель, который знает как дополнять класс.
     */
    fun bind(clz: Class<*>, completer: CommandCompleter) {
        completers[clz] = completer
    }

}
